### DISCUSSION TITLE
What is the Prefect solution for airflow's KubernetesPodOperator?

### DISCUSSION THREAD

**Discussion by pietermarsman** [PrefectHQ/prefect#3931] [Q&A] [answered]
Hi there!

I'm considering Prefect as a possibility to run our ml training workflows with. The training of our image processing models consists of several steps, from downloading data, to training models, to evaluating and deploying models. The models that we use are sometimes implemented in tensorflow and sometimes in pytorch. For that reason I like to keep the dependencies of tasks strictly separated. 

Additionally, I would like to schedule these trainings once per month or so and run everything on kubernetes. 

With Airflow I noticed I can use the [KubernetesPodOperator](https://incubator-airflow.readthedocs.io/en/latest/howto/operator/kubernetes.html). This runs a operator (task in prefect lingo) as a pod on a kubernetes cluster. I can wrap all the training code for the different models into independent docker images that run the training and exit at the end. 

What is the Prefect way of doing this? 

I've considered:

- using the [KubernetesAgent](https://docs.prefect.io/orchestration/agents/kubernetes.html). But I get that this schedules entire flows in a single docker image that runs as a pod on k8s. I want to specify the docker image for each of the tasks in a flow. 
- using [BuildImage](https://docs.prefect.io/api/latest/tasks/docker.html#buildimage) and [StartContainer](https://docs.prefect.io/api/latest/tasks/docker.html#startcontainer) to pull and run an image. But this (I guess) runs the docker on the host. I want to run it as a pod. 
- using [ RunNamespacedJob](https://docs.prefect.io/api/latest/tasks/kubernetes.html#runnamespacedjob). This looks the most promising, but I'm unsure about the amount of manual configuration I need to do. It looks like I need to write the entire k8s yaml to specify the pod. E.g. which volumes to mount. 

Any tips and tricks on how I should approach this are appreciated.

**Comment by pietermarsman** [comment, marked as answer]
I guess I'm asking about this feature: https://youtu.be/TlawR_gi8-Y?t=1978

**Reply by pietermarsman** [reply]
This looks also related: https://stackoverflow.com/questions/62509406/docker-run-as-prefect-task

**Reply by joshmeek** [reply]
Hi @pietermarsman, yes the `RunNamespacedJob / CreateNamespacedPod` tasks would be what you are after. Currently they are set to accept kubernetes yaml manifests but if you think that there is a better way to provide these options the task library is always accepting community contributions!
